decompressed/base/lib/mount_modroot/04_move_overlay

```diff
@@ -0,0 +1,22 @@
+#!/bin/sh
+
+upperdir=$1
+overlay_bank=/overlay/bank_2
+
+#use the www folder as check to know that we should populate the modoverlay (mtd3) and cleanup real overlay (mtd2)
+if [ -d $overlay_bank/www ]; then
+  move_files_and_clean(){
+    for file in $(find "$1"*/ -xdev | cut -d '/' -f4-); do
```

Comment:
Ma questo ciclo fatto così avete verificato che non cancelli anche I mountpoint dell'overlay superiore?

Reply:
sposta tutto quello che trova in /overlay/bank_2 e quando ha finito ci fa pure un rm -rf /overlay/bank_2/* sicuro c'è qualche errore stupido in tutto ciò che poi mi porta ad averlo dopo la prova di riavvio con solo quelle cartelle di mnt e saferoot

Non hai capito perchè ho scritto la cartella sbagliata..
il punto è che anche sistemando il tutto, con in /overlay/bank_2 solo quei file che abbiamo messo in lista vanno a farsi benedire anche upgrade e reset (questo no usando l'rtfd attuale delal gui) perchè gli script in `mount_root` ovviamente quando eseguiti non troveranno l'overlay di mtd3

Reply:
Nel momento in cu vai ad eseguire questo script, prima del pivoting, esiste una cartella vuota /modoverlay (quindi scritta in /overlay/bank_2/modoverlay) sulla quale mtd3 è stata montata. Questa cartella vuota (così come appare in /overlay/bank_2) finisce per essere copiata in mtd3 e rimossa da mtd2. A questo punto succedono cose molto brutte che non so cosa possano provocare:
- la directory /modoverlay che in quel momento fa da mountpoint di mtd3 in realtà non esiste più da nessuna parte
- nel momento in cui il nuovo overlay nasce in /mnt/modroot come sovrapposizione di /modoverlay e / (cioè la immediatamente futura /saferoot), la prima, cioè l'upper layer, in realtà non esiste, o non so in che strana condizione sia, ma essendoci in mtd3 la copia della cartella vuota in origine situata in mtd2, nel risultato si ha che la nuova /mnt/modroot ha al suo interno una /mnt/modroot/modoverlay
- sul finire del pivoting, il mountpoint di mtd3 (che in quel momento risulta montato su una directory che non esiste più viene spostato su una che invece esiste cioè quella /mnt/modroot/modoverlay oramai divenuta /modoverlay
C'è quindi un lasso di tempo, successivo alla copia/eliminazione dei file da mtd2 a mtd3, e precedente la nascita di /mnt/modroot durante il quale secondo me /modoverlay è un mezzo cadavere nel quale chissà come sei riuscito a continuare a scrivere dopo avergliela tolta da sotto il sedere, ma nel quale probabilmente non riesci a leggere i file che intendi ricopiare indietro in mtd2. Devi dunque provare a vedere se risolvi escludendo /overlay/bank_2/modverlay da quel ciclo di copia/rimozione

Dopo l'esecuzione dello script in mount_modroot, avviene la creazione della cartella /saferoot (quindi /overlay/bank_2/saferoot) e di altre in /mnt (quindi /overlay/bank_2/mnt) necessarie al mount del nuovo overlay. Quindi è normale che tu veda in /overlay/bank_2 /saferoot e /mnt

Reply:
Il fatto che invece lo script di preserve non venga eseguito è ovvio, il ciclo for che li esegue uno per volta quando prova ad eseguirlo non lo trova perchè hai appena spostato anche lui.

Reply:
si risolve fondendoli in un singolo file

Reply:
ok dovrebbe essermi tutto chiaro, e il funzionamento "magico" è spiegato dal fatto che quel loop non elimina le cartelle ma solo i file

le cartelle vengono fatte fuori con l' `rm -rf "$1"*` che a questo punto causa anche la sparizione del mountpoint /modoverlay quando poi lo script 05 viene eseguito non trova /modoverlay e quindi non riesce a copiare dietro un ciufolo

a questo punto mi resta il piccolo dubbio di capire chi ricrea /modoverlay e /mnt .. immagino il pivoting stesso, e non credo se il ricreare il mountpoint subito dopo il remove possa essere una soluzione veloce senza sbattermi oltre

Reply:
> perchè hai appena spostato anche lui

giusto, mindfuck.

Reply:
> a questo punto mi resta il piccolo dubbio di capire chi ricrea /modoverlay e /mnt ..

la /modoverlay usata alla fine del giro è la copia di quella originale, ma è chiaro che successive esecuzioni di quello script la ricreano non trovandola anche quando poi il transfer non avviene
/mnt e /saferoot invece le ricrea il pivoting

Reply:
> e non so se il ricreare il mountpoint subito dopo il remove possa essere una soluzione veloce senza sbattermi oltre

No, potrebbe peggiorare ulteriormente le cose. Come hai giustamente detto l'rm lo fa sul finale, quindi in realtà non scrivi mai più in mtd3 dopo avergli segato il mountpoint, forse potrebbe non essere necessario intervenire.

Altrimenti puoi sistemare entrambe le cose escludendo i file in preserve_list dal ciclo.

Reply:
a questo punto dato che nella pratica è inutilizzabile non so quanto intelligente sia mantenere la cartella `mount_modroot` (e probabilmente anche `mount_modroot_checks`) non converrebbe spostare tutto (anche lo spostamento dei file) nel file `mount_root-mod` ?

la cartella /mnt ha bisogno anche lei di cure particolari?

Reply:
non è inutilizzabile, purchè transfer e preserve siano nel medesimo file qualsiasi altro script precedente o successivo ma correttamente preservato rispettano il design di quella logica

quelli di modroot_checks non soffrono di problemi analoghi perchè vengono tutti eseguiti prima

nessuna cura particolare per /mnt

Reply:
non avrebbe comunque senso farlo in mount_root-mod perchè se a romperne la logica è uno degli script che include sta agli script che include evitare che accada

Reply:
anche lo stesso mount dell'overlay fatto in preinit segue la stessa logica, se lo script inserito in preinit incaricato di farlo rompesse il preinit non sarebbe corretto andare a gestire direttamente il mount dell'overlay nello script che orchestra il preinit

Reply:
dai un occhio ora prima che gli faccio fare un'altro giro :D